### PR TITLE
Add periodic etcd scraping to integration tests

### PR DESCRIPTION
**Pull request by MikeSpreitzer**
#### What type of PR is this?

/kind feature
/kind flake

#### What this PR does / why we need it:
This PR adds periodic Prometheus scraping of the etcd server used in integration testing.
This is helpful because we suspect integration tests are flaky due to etcd overload.
See https://github.com/kubernetes/kubernetes/issues/106012#issuecomment-961953255

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
Integration testing now takes periodic Prometheus scrapes from the etcd server.
There is a new script ,`hack/run-prometheus-on-etcd-scrapes.sh`, that runs a containerized Prometheus server against an archive of such scrapes.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

/sig testing
/sig api-machinery
/cc @aojea 
@wojtek-t 
@tkashem

### CODE REVIEW

**Comment by MikeSpreitzer** [comment]
A possible alternative approach would be to run the regular Prometheus server to do the scraping.  In order to draft that, I would have to understand the best way to run that server.  I can think of two approaches, each of which involves things I do not know.

1. Run Prometheus in a container.  What --- if any --- is the right way to add running containers to the integration tests?  BTW, https://github.com/kubernetes/community/pull/6203 begs a related question.
2. Build and install Prometheus as part of running the integration tests.  https://github.com/prometheus/prometheus#building-from-source says this requires npm >= v7 and NodeJS >= v16; are these dependencies already satisfied or reasonable to introduce?

**Comment by aojea** [comment]
there are some shellcheck errrors

> > In ./hack/lib/etcd.sh line 120:
>     tar czf "${ETCD_SCRAPE_DIR}/scrapes.tgz" "${ETCD_SCRAPE_DIR}"/*.scrape && rm "${ETCD_SCRAPE_DIR}"/*.scrape || :
>                                                                            ^-- SC2015: Note that A && B || C is not if-then-else. C may run when A is true.
> 
> 
> In ./hack/make-rules/test-integration.sh line 67:
>   local ETCD_SCRAPE_PID
>         ^-------------^ SC2034: ETCD_SCRAPE_PID appears unused. Verify use (or export if used externally).

 metrics are obtained in https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/106190/pull-kubernetes-integration/1456747488651251712/artifacts/etcd-scrapes/scrapes.tgz , maybe we can compress without the full path?
right now it expands to:
> ./logs/artifacts/etcd-scrapes/224457.scrape

naive question, how can I visualize those dumps , is there any tool to import those files into prometheus locally? @serathius  do you know?

/assign @thockin  @BenTheElder 
for bash review

**Comment by MikeSpreitzer** [comment]
Revised to make the scrapes easier to handle, and added a simple server for scrape files that exports their contents via the usual Prometheus web interface.  Two easy missing pieces: (1) something that unpacks the tarchive and invokes the simple server, and (2) an appropriate Prometheus config and a script to invoke Prometheus on that.

**Comment by aojea** [comment]
this works fine, awesome job, Mike, just the netcat flag comment

For docs purposes, you can also use grafana dashboards to check the metrics:

1. run another container with grafana
```
docker run -p 3000:3000 grafana/grafana
```

2. login with admin:admin on `localhost:3000`
3. setup local prometheus as data source
4. install an etcd dashboard, (id 3070 `Etcd by Prometheus`) per example
5. Check all the metrics of etcd during the test
image

**Comment by MikeSpreitzer** [comment]
Speaking of doc, I was wondering where is the right place to document the new scripts. They are non-trivial, and deserve a bit more documentation than the brief release notes mention that I drafted.

File: hack/serve-prom-scrapes.sh [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+find_and_transform() {
+    echo -n $'HTTP/1.0 200 OK\r\nContent-Type: text/plain\r\n\r\n' > "$response_file"
+    find "$scrapes_dir" -name "*.scrape" -print0 | sort -z | while read -d '' -r scrapename; do transform "$scrapename" >> "$response_file"; done
+}
+
+find_and_transform
+
+while true; do nc -l -N -p "$port_num" < "$response_file" > /dev/null; sleep 10; done
```
[the remark this answers is not in the document]
Comment (aojea) [reply hack/serve-prom-scrapes.sh:59]: I think `nc -l -p` is the most universal combination of flags

[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply hack/serve-prom-scrapes.sh:59]: Oh, that's annoying.  It is missing in MacOS too.  If I just remove the `-N`, this script does not work on MacOS.  Does it work for you?  What OS are you using?  Which netcat package?  I got this working on Ubuntu 20.04 using netcat-openbsd 1.206-1ubuntu1.  Between that and MacOS, I can tell which is which using `man nc | grep -w -E -N`; what does that get for you?

[the remark this answers is not in the document]
Comment (aojea) [reply hack/serve-prom-scrapes.sh:59]: oh, so this is the problem
```
     -N      shutdown(2) the network socket after EOF on the input.  Some servers require this to finish their work.
```

then keep the -N , is the one implemented in the original netcat

**Comment by MikeSpreitzer** [comment]
The force-push to c00d4c3800f adds some more flexibility to the scripts.
Now they work on my Mac, albeit with the _occasional_ mysterious print of the scrape request message from Prometheus.

**Comment by MikeSpreitzer** [comment]
The last few force-pushes have been tweaking up the log message(s) about networking.

**Comment by MikeSpreitzer** [comment]
@deads2k 
@tkashem

File: hack/make-rules/test-integration.sh [hunk 2]
```diff
@@ -64,6 +64,9 @@ runTests() {
   kube::log::status "Starting etcd instance"
   CLEANUP_REQUIRED=1
   kube::etcd::start
+  # shellcheck disable=SC2034
+  local ETCD_SCRAPE_PID # Set in kube::etcd::start_scraping, used in cleanup
+  kube::etcd::start_scraping
```
Comment (deads2k) [inline hack/make-rules/test-integration.sh:69]: The idea of doing this scrape using a simple `curl` for integration tests seems lightweight enough to try.  I'm not entirely sure how you will stitch this back together with which tests were running at a particular time, but given that it doesn't really add to our dependency stack, I'm ok giving it a try.

Comment (aojea) [reply hack/make-rules/test-integration.sh:69]: the job stores the etcd logs and the test output with timestamps, it is not great but better than before :shrug:

Comment (MikeSpreitzer) [reply hack/make-rules/test-integration.sh:69]: In the case I was looking at, there was a log with timestamps.

**Comment by MikeSpreitzer** [comment]
The force-push to 4ca4ccdcc21 picks some shell lint and tweaks logging again.

**Comment by deads2k** [comment]
a lightweight way to keep the etc metrics for integration tests seems worth trying.

/approve

**Pull request merged**